source/grid/manifold_lib.cc

```diff
@@ -1047,18 +1061,34 @@ namespace
         cell.vertex(4), cell.vertex(5), cell.vertex(6), cell.vertex(7)
       }
     };
-    Point<spacedim> new_point;
 
+    // store the components of the linear shape functions because we need them
+    // repeatedly. we allow for 10 such shape functions to wrap around the
+    // first four once again for easier face access.
+    double linear_shapes[10];
+    for (unsigned int d=0; d<3; ++d)
+      {
+        linear_shapes[2*d]   = 1.-chart_point[d];
+        linear_shapes[2*d+1] = chart_point[d];
+      }
+
+    Point<spacedim> new_point;
     if (cell_is_flat)
-      for (unsigned int v=0; v<GeometryInfo<3>::vertices_per_cell; ++v)
-        new_point += GeometryInfo<3>::d_linear_shape_function(chart_point, v) *
-                     vertices[v];
+      for (unsigned int i2=0, v=0; i2<2; ++i2)
+        for (unsigned int i1=0; i1<2; ++i1)
+          for (unsigned int i0=0; i0<2; ++i0, ++v)
+            new_point += (linear_shapes[4+i2] * linear_shapes[2+i1]) *
+                         linear_shapes[i0] * vertices[v];
     else
       {
         // identify the weights for the vertices and lines to be accumulated
         double weights_vertices[GeometryInfo<3>::vertices_per_cell];
-        for (unsigned int v=0; v<GeometryInfo<3>::vertices_per_cell; ++v)
-          weights_vertices[v] = GeometryInfo<3>::d_linear_shape_function(chart_point, v);
+        for (unsigned int i2=0, v=0; i2<2; ++i2)
+          for (unsigned int i1=0; i1<2; ++i1)
+            for (unsigned int i0=0; i0<2; ++i0, ++v)
+              weights_vertices[v] = (linear_shapes[4+i2] * linear_shapes[2+i1]) *
+                                    linear_shapes[i0];
```

Comment:
We could do a similar trick here (move this above the `new_point` calculation and reuse the values)